<a href="https://colab.research.google.com/github/Abdulrahman2k/mlcourse_dubai/blob/master/topic_1_exploratory_data_analysis_with_pandas_ASSIGNMENT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center>
<img src="https://github.com/Abdulrahman2k/mlcourse_dubai/blob/master/img/ods_stickers.jpg?raw=1">
    
## [mlcourse.ai](https://mlcourse.ai) - Open Machine Learning Course

Author: [Yury Kashnitsky](https://www.linkedin.com/in/festline/). Translated and edited by [Sergey Isaev](https://www.linkedin.com/in/isvforall/), [Artem Trunov](https://www.linkedin.com/in/datamove/), [Anastasia Manokhina](https://www.linkedin.com/in/anastasiamanokhina/), and [Yuanyuan Pao](https://www.linkedin.com/in/yuanyuanpao/). All content is distributed under the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

# <center> Assignment #1 (demo)
## <center>  Exploratory data analysis with Pandas


**Same assignment as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/a1-demo-pandas-and-uci-adult-dataset) + [solution](https://www.kaggle.com/kashnitsky/a1-demo-pandas-and-uci-adult-dataset-solution).**

**In this task you should use Pandas to answer a few questions about the [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) dataset. (You don't have to download the data – it's already  in the repository). Choose the answers in the [web-form](https://docs.google.com/forms/d/1uY7MpI2trKx6FLWZte0uVh3ULV4Cm_tDud0VDFGCOKg).**

Unique values of all features (for more information, please see the links above):
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [0]:
import numpy as np
import pandas as pd
pd.set_option('display.max.columns', 100)
# to draw pictures in jupyter notebook
%matplotlib inline 
import matplotlib.pyplot as plt
import seaborn as sns
# we don't like warnings
# you can comment the following 2 lines if you'd like to
import warnings
warnings.filterwarnings('ignore')

In [0]:
data = pd.read_csv('https://raw.githubusercontent.com/Abdulrahman2k/mlcourse_dubai/master/data/adult.data.csv')
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. How many men and women (*sex* feature) are represented in this dataset?** 

In [0]:
data["sex"].value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

**2. What is the average age (*age* feature) of women?**

In [0]:
data[data["sex"]=="Female"]["age"].mean()

36.85823043357163

**3. What is the percentage of German citizens (*native-country* feature)?**

In [0]:
(data[data["native-country"]=="Germany"]["race"].count()/data["race"].count())*100
#data["race"].value_counts()

0.42074874850281013

**4-5. What are the mean and standard deviation of age for those who earn more than 50K per year (*salary* feature) and those who earn less than 50K per year? **

In [0]:
columns_to_show = ['age']


data.groupby(['salary'])[columns_to_show].agg([np.mean, np.std])

age           
             mean        std
salary                      
<=50K   36.783738  14.020088
>50K    44.249841  10.519028

**6. Is it true that people who earn more than 50K have at least high school education? (*education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate* feature)**

In [0]:
data[data['salary']=='>50K']['education'].value_counts()
# as per data people earning more than 50K can come from any education background and do not have at least high school education

Bachelors       2221
HS-grad         1675
Some-college    1387
Masters          959
Prof-school      423
Assoc-voc        361
Doctorate        306
Assoc-acdm       265
10th              62
11th              60
7th-8th           40
12th              33
9th               27
5th-6th           16
1st-4th            6
Name: education, dtype: int64

## As per data people earning more than 50K can come from any education background and do not have at least high school education, so the answer is that the assumption that people earning more than 50k has at least high school eduction is **NOT** True

**7. Display age statistics for each race (*race* feature) and each gender (*sex* feature). Use *groupby()* and *describe()*. Find the maximum age of men of *Amer-Indian-Eskimo* race.**

In [0]:
columns_to_group =   ['race','sex']

columns_to_show =['age']


data.groupby(columns_to_group)[columns_to_show].agg([np.mean, np.std,np.min, np.max])

##the maximum age of men of Amer-Indian-Eskimo race.**

age                     
                                mean        std amin amax
race               sex                                   
Amer-Indian-Eskimo Female  37.117647  13.114991   17   80
                   Male    37.208333  12.049563   17   82
Asian-Pac-Islander Female  35.089595  12.300845   17   75
                   Male    39.073593  12.883944   18   90
Black              Female  37.854019  12.637197   17   90
                   Male    37.682600  12.882612   17   90
Other              Female  31.678899  11.631599   17   74
                   Male    34.654321  11.355531   17   77
White              Female  36.811618  14.329093   17   90
                   Male    39.652498  13.436029   17   90

## The maximum age of men of Amer-Indian-Eskimo race is **82** Years 

**8. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (*marital-status* feature)? Consider as married those who have a *marital-status* starting with *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.**

In [0]:
Married_function =lambda maritalstatus:maritalstatus.startswith('Married')
Unmarried_function =lambda maritalstatus:not(maritalstatus.startswith('Married'))
Greater_earning = lambda earning:earning=='>50K'


In [0]:
##data[data['marital-status'].str.slice(0,7)=='Married']
round( (data[data['marital-status'].apply(Married_function) & data['salary'].apply(Greater_earning)]['age'].count() / data['age'].count()) * 100)

21.0

In [0]:

round( (data[data['marital-status'].apply(Unmarried_function) & data['salary'].apply(Greater_earning)]['age'].count() / data['age'].count()) * 100)

3.0

### As per Data Around 21 Percent of All Married Personnel earn more than 50K  and only 3 percent of Unmarried Personnel earn more than 50K

In [0]:
columns_to_group =   ['race','sex']

columns_to_show =['age']


data.groupby(columns_to_group)[columns_to_show].agg([np.mean, np.std,np.min, np.max])

**9. What is the maximum number of hours a person works per week (*hours-per-week* feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?**

In [0]:
# You code here
np.max(data['hours-per-week'])

99

### The Maximum hours per week worked is **99** Hours

In [0]:
data[data['hours-per-week']==np.max(data['hours-per-week'])]['age'].count()

85

### The Number of Personnel working the Maximum Number of hours (**99** Hours) worked per week is **85** Persons

In [0]:
 round((data[data['hours-per-week']==np.max(data['hours-per-week']) & data['salary'].apply(Greater_earning) ]['age'].count() /data[data['hours-per-week']==np.max(data['hours-per-week'])  ]['age'].count() ) *100,2)

2.35

### The percentage of personnel earning more than 50K and working maximum number of hours per week, compared to all personnel working maximum number of hours per week is **2.35** Percentage 

**10. Count the average time of work (*hours-per-week*) for those who earn a little and a lot (*salary*) for each country (*native-country*). What will these be for Japan?**

In [0]:
columns_to_group =   ['salary','native-country']

columns_to_show =['hours-per-week']


data.groupby(columns_to_group)[columns_to_show].agg([np.mean])



hours-per-week
                                 mean
salary native-country                
<=50K  ?                    40.164760
       Cambodia             41.416667
       Canada               37.914634
       China                37.381818
       Columbia             38.684211
...                               ...
>50K   Thailand             58.333333
       Trinadad&Tobago      40.000000
       United-States        45.505369
       Vietnam              39.200000
       Yugoslavia           49.500000

[82 rows x 1 columns]

In [0]:
data[data['native-country']=='Japan'].groupby(columns_to_group)[columns_to_show].agg([np.mean])

hours-per-week
                                mean
salary native-country               
<=50K  Japan               41.000000
>50K   Japan               47.958333

### The average number of hours for personnel earning more than 50K is **41.00** hours and the average number of hours per week for personnel earning more than 50k is **47.96** hours per week